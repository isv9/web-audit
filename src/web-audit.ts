import { auditBlockSemantics, auditCommonSemantics, auditHeaderSemantics, auditTextSemantics } from './audits/semantic'
import { auditLinks } from './audits/links'

export type WebDocument = {
  getElementsByTagNameCount (tag: string): number;
  querySelectorAll (query: string): { length: number };
  getEmptyElementsByTagName (tag: string): { length: number }
};

type AuditSectionResult = {
  name: string;
  auditResults: AuditResult[];
};

export type AuditResult = {
  name?: string;
  tables: (AuditResultTable)[];
  liveCollections?: (AuditResultLiveCollection)[];
  logs?: string[];
  warnings?: string[];
  errors?: string[];
};

type AuditResultTable = { name?: string } & { [key: string]: string | number };
export type AuditResultLiveCollection = { name: string, collection: object };

export class WebAudit {

  private readonly webDocument: WebDocument

  constructor (webDocument: WebDocument) {
    this.webDocument = webDocument
  }

  audit () {
    const auditSemanticsResult = this.auditSemantics()
    const auditLinksResult = this.auditLinks()
    console.group('web audit')
    WebAudit.renderAuditSectionResult(auditSemanticsResult)
    WebAudit.renderAuditSectionResult(auditLinksResult)
    console.groupEnd()
  }

  private auditLinks (): AuditSectionResult {
    return {
      name: 'links',
      auditResults: [
        auditLinks(this.webDocument)]
    }
  }

  private auditSemantics (): AuditSectionResult {

    return {
      name: 'semantics',
      auditResults: [auditCommonSemantics(this.webDocument),
        auditHeaderSemantics(this.webDocument),
        auditTextSemantics(this.webDocument),
        auditBlockSemantics(this.webDocument)]
    }
  }

  static renderAuditSectionResult (auditSectionResult: AuditSectionResult) {
    const { name, auditResults } = auditSectionResult
    const details = {
      errorsCount: 0,
      warningsCount: 0
    }
    auditResults.forEach(auditResult => {
      const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
      const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
      details.errorsCount += errorsCount
      details.warningsCount += warningsCount
    })
    console.group(`${name}, (errors="${details.errorsCount}", warnings="${details.warningsCount}")`)
    auditResults.forEach(auditResult => WebAudit.renderAuditResult(auditResult))
    console.groupEnd()
  }

  static renderAuditResult (auditResult: AuditResult) {
    const { name, tables, errors = [], warnings = [], liveCollections = [], logs = [] } = auditResult
    const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
    const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
    console.group(`${name}, (errors="${errorsCount}", warnings="${warningsCount}")`)
    tables.forEach(({ name, nodes, ...tableDataSet }) => {
      console.log(name)
      console.table(tableDataSet)
    })
    liveCollections.forEach(({ name, collection }) => {
      console.log(name)
      console.log(collection)
    })
    logs.forEach(message => console.log(message))
    console.log('Summary')
    if (errorsCount > 0 || warningsCount > 0) {
      errors.forEach(message => console.error(message))
      warnings.forEach(message => console.warn(message))
    } else {
      console.log('It is OK')
    }

    console.groupEnd()
  }

  static getAuditResultMessagesCount (auditResult: AuditResult, key: keyof AuditResult): number {
    let property = auditResult[key]
    if (property === null || property === undefined) {
      property = []
    }
    if (Array.isArray(property)) {
      return property.length
    }
    throw new Error('Property type must be array')
  }
}

