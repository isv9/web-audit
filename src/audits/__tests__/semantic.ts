import { WebDocument } from '../../web-audit';
import {
  auditBlockSemantics,
  auditCommonSemantics,
  auditHeaderSemantics,
  auditTextSemantics,
} from '../semantic';

describe('auditCommonSemantics', () => {
  it('document has each common elements', () => {
    const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(() => 1),
    };
    expect(auditCommonSemantics(document)).toMatchSnapshot();
  });
  it('document has two main tags', () => {
    const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(tag => (tag === 'main' ? 2 : 1)),
    };
    expect(auditCommonSemantics(document)).toMatchSnapshot();
  });

  it('document does not have any common tag', () => {
    const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(() => 0),
    };
    expect(auditCommonSemantics(document)).toMatchSnapshot();
  });
});

it('auditTextSemantics', () => {
  let inc = 0;
  const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
    getElementsByTagNameCount: jest.fn(() => ++inc),
  };
  expect(auditTextSemantics(document)).toMatchSnapshot();
});

it('document does not any text tag', () => {
  const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
    getElementsByTagNameCount: jest.fn(() => 0),
  };
  expect(auditTextSemantics(document)).toMatchSnapshot();
});

describe('auditHeaderSemantics', () => {
  it('document has some headers', () => {
    let inc = 0;
    const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(() => ++inc),
    };
    expect(auditHeaderSemantics(document)).toMatchSnapshot();
  });
  it('document does not have any headers', () => {
    const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(() => 0),
    };
    expect(auditHeaderSemantics(document)).toMatchSnapshot();
  });
  it('document has two h1', () => {
    const document: Pick<WebDocument, 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(tag => (tag === 'h1' ? 2 : 0)),
    };
    expect(auditHeaderSemantics(document)).toMatchSnapshot();
  });
});

describe('auditBlocksSemantics', () => {
  it('document has good block semantic', () => {
    const document: Pick<WebDocument, 'getEmptyElementsByTagName' | 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(() => 20),
      getEmptyElementsByTagName: jest.fn(() => ({ length: 0 })),
    };
    expect(auditBlockSemantics(document)).toMatchSnapshot();
  });

  it('document has divatos and empty articles', () => {
    const document: Pick<WebDocument, 'getEmptyElementsByTagName' | 'getElementsByTagNameCount'> = {
      getElementsByTagNameCount: jest.fn(tag => (tag === 'div' ? 100 : 20)),
      getEmptyElementsByTagName: jest.fn(tag => ({ length: tag === 'article' ? 2 : 0 })),
    };
    expect(auditBlockSemantics(document)).toMatchSnapshot();
  });
});
