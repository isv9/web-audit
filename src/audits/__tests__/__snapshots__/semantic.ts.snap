// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditBlocksSemantics document has divatos and empty articles 1`] = `
Object {
  "liveCollections": Array [
    Object {
      "collection": Object {
        "length": 2,
      },
      "name": "Empty \\"article\\" elements count in dom",
    },
  ],
  "name": "block elements count in dom",
  "tables": Array [
    Object {
      "address": 20,
      "article": 20,
      "button": 20,
      "details": 20,
      "div": 100,
      "form": 20,
      "li": 20,
      "section": 20,
      "summary": 20,
      "ul": 20,
    },
  ],
  "warnings": Array [
    "Document has much div",
    "Document has empty blocks",
  ],
}
`;

exports[`auditBlocksSemantics document has good block semantic 1`] = `
Object {
  "liveCollections": Array [],
  "name": "block elements count in dom",
  "tables": Array [
    Object {
      "address": 20,
      "article": 20,
      "button": 20,
      "details": 20,
      "div": 20,
      "form": 20,
      "li": 20,
      "section": 20,
      "summary": 20,
      "ul": 20,
    },
  ],
  "warnings": Array [],
}
`;

exports[`auditCommonSemantics document does not have any common tag 1`] = `
Object {
  "errors": Array [],
  "name": "common elements count in dom",
  "tables": Array [
    Object {
      "footer": 0,
      "header": 0,
      "main": 0,
      "nav": 0,
    },
  ],
  "warnings": Array [
    "Document does not have main tag",
    "Document does not have header tag",
  ],
}
`;

exports[`auditCommonSemantics document has each common elements 1`] = `
Object {
  "errors": Array [],
  "name": "common elements count in dom",
  "tables": Array [
    Object {
      "footer": 1,
      "header": 1,
      "main": 1,
      "nav": 1,
    },
  ],
  "warnings": Array [],
}
`;

exports[`auditCommonSemantics document has two header tags 1`] = `
Object {
  "errors": Array [],
  "name": "common elements count in dom",
  "tables": Array [
    Object {
      "footer": 1,
      "header": 2,
      "main": 1,
      "nav": 1,
    },
  ],
  "warnings": Array [],
}
`;

exports[`auditCommonSemantics document has two main tags 1`] = `
Object {
  "errors": Array [
    "Document must have one main tag",
  ],
  "name": "common elements count in dom",
  "tables": Array [
    Object {
      "footer": 1,
      "header": 1,
      "main": 2,
      "nav": 1,
    },
  ],
  "warnings": Array [],
}
`;

exports[`auditHeaderSemantics document does not have any headers 1`] = `
Object {
  "errors": Array [
    "Document must have one h1",
  ],
  "name": "header elements count in dom",
  "tables": Array [
    Object {
      "h1": 0,
      "h2": 0,
      "h3": 0,
      "h4": 0,
      "h5": 0,
      "h6": 0,
    },
  ],
}
`;

exports[`auditHeaderSemantics document has some headers 1`] = `
Object {
  "errors": Array [],
  "name": "header elements count in dom",
  "tables": Array [
    Object {
      "h1": 1,
      "h2": 2,
      "h3": 3,
      "h4": 4,
      "h5": 5,
      "h6": 6,
    },
  ],
}
`;

exports[`auditHeaderSemantics document has two h1 1`] = `
Object {
  "errors": Array [
    "Document must have one h1",
  ],
  "name": "header elements count in dom",
  "tables": Array [
    Object {
      "h1": 2,
      "h2": 0,
      "h3": 0,
      "h4": 0,
      "h5": 0,
      "h6": 0,
    },
  ],
}
`;

exports[`auditTextSemantics 1`] = `
Object {
  "errors": Array [],
  "name": "text elements count in dom",
  "tables": Array [
    Object {
      "blockquote": 1,
      "dd": 2,
      "dl": 3,
      "dt": 4,
      "name": "text elements",
      "pre": 5,
    },
    Object {
      "abbr": 15,
      "b": 8,
      "cite": 16,
      "code": 11,
      "em": 14,
      "i": 9,
      "name": "text semantic elements",
      "p": 7,
      "small": 10,
      "span": 6,
      "strong": 12,
      "time": 13,
    },
  ],
}
`;

exports[`document does not any text tag 1`] = `
Object {
  "errors": Array [
    "Document does not any text tag",
  ],
  "name": "text elements count in dom",
  "tables": Array [
    Object {
      "blockquote": 0,
      "dd": 0,
      "dl": 0,
      "dt": 0,
      "name": "text elements",
      "pre": 0,
    },
    Object {
      "abbr": 0,
      "b": 0,
      "cite": 0,
      "code": 0,
      "em": 0,
      "i": 0,
      "name": "text semantic elements",
      "p": 0,
      "small": 0,
      "span": 0,
      "strong": 0,
      "time": 0,
    },
  ],
}
`;
